### ISSUE TITLE
Unable to ingest Aster from HDFS

### ISSUE THREAD

**Issue by wncampbell** [ngageoint/mrgeo#61]
Multiple attempts to ingest aster from HDFS have resulted in "Invocation target error". I've tried the ingest with a memoryintensive.multiplier value of 2,8,12, and 32. The raw aster was copied to hdfs using distcp and I verified that all data copied over correctly.

The error happens seemingly randomly, sometimes failing after 15min, sometimes not for an hour. The mrgeo command is:

`mrgeo ingest -nd -9999 -sk -sp -z 12 -o /mrgeo/images/aster-hdfs-30m /mrgeo/raw-images/aster-30m`

Ingest from S3 runs successfully with a multiplier of 12.

**Comment by wncampbell** [comment]
At the suggestion of @djohnson729 I've reduced dfs replication down to 1 from the default of 3:

`Path=s3://elasticmapreduce/bootstrap-actions/configure-hadoop,Name="Reduce HDFS replication",Args=["-h","dfs.replication=1"]`

**Comment by djohnson729** [comment]
Reducing replication to 1 is to determine if that helps with the the "No space of device" errors that I saw in the logs from Nathan's earlier failed jobs. However, using that replication value, it could possibly make some jobs run slower since it can limit hadoop's options for data locality.

**Comment by wncampbell** [comment]
With replication at 1 the ingest from HDFS still fails. HDFS shows only 26% disk usage with 5.3 TB still available when the job failed.

**Comment by wncampbell** [comment]
Logs do have the "no space left on device" error but hdfs shows plenty of free space.

**Comment by wncampbell** [comment]
Doubled the node count (hdfs size now 14TB) and same failure. Ran a small ingest on aster-30m-xsmall successfully.

**Comment by ttislerdg** [comment]
I just finished investigating this and captured much more information.  It is indeed "no space left on device", but it's not what you might think.  

With 100 m3.xlarge nodes, there is about 7.1T on the mounted drives.  It ends up this space is used for both HDFS *and* hadoop/spark temp space.

HDFS:  source imagery:  550G (~1.6T replicated),  ingested imagery (no pyramids) ~1.4T (~4.2T replicated)

So just storing aster source and the base level takes ~ 5.8T, leaving 2.3T for temp space.  It ends up that spark will need at minimum as much temp space as the ingested imagery, since it will spill to disk if needed for shuffling and persisting..

As can be expected, it is during the save phase the cluster runs out of memory, that's when spark has the data cached on disk and we start to write to HDFS.

So the solution is to make sure the cluster has enough disk space to accommodate HDFS *and* temp space.  I started a 200 node cluster, with 14T space, and it was enough to complete successfully.  I ran the ingest multiple times and it never failed. 

Here's some interesting stats I collected on one data node in the cluster showing how tmp and DFS space were being used:

phase | tmp | hdfs | description
------------ | ------------ | ------------- | -------------
start | 400m | 4.9G | beginning of run
flatmap | :arrow_up: 2.5G | :arrow_right: 4.9G | tile individual input files into TMS tiles 
aggregate | :arrow_up: 4.9G | :arrow_right: 4.9G  | calculate input bounds
reducebykey | :arrow_up: 7.4G :arrow_down: 6.5G | :arrow_right: 4.9G | repartition data to output rows
save | :arrow_right: 7.4G | :arrow_up: 17G | sort within partitions and save to HDFS
collect | :arrow_up: 7.6G | :arrow_right: 17G | calculate "splits" file (start and end tileIds) for each partition
aggregate | :arrow_right: 7.6G | :arrow_right: 17G | calculate statistics
end | :arrow_down: 400m | :arrow_right: 17G | end of run
